torch/distributed/pipelining/schedules.py

```diff
@@ -743,6 +753,99 @@ def _step_microbatches(
+- Should '_Action' enum contain all the comm stuff too? how do we separate the basic actions from the full set?
+- Where should 'add_comms' be called? should it be called by a particular Schedule class's __init__, and possibly
+    pass flags to control the heuristics of add_comms based on that schedule?
+
+TODO
+- add a few more test cases for the shard/unshard part
+  - refactor test helper for shard/unshard to only worry about one rank at a time
+- then do send/recv part and simplify the runtime. set up e2e tests.
+"""
+
+
+def _add_unshard_reshard(
+    compute_actions: List[Optional[_Action]],
+    max_active_stages: int = 3,
+) -> List[Optional[_Action]]:
+    """
+    Adds unshard/reshard actions to the schedule.
+
+    we abandon the "timestep lock"  during lowering
+
+    max_active_stages controls how many prefetches we allow. It should be measured in mb and tuneable but in practice
+    3 stages is probably the thing we want?
+    (to account for having one f and one b active, and something else prefetching?)
+    """
+
+    def next_stage_indices(
+        count: int, next_actions: List[Optional[_Action]]
+    ) -> List[int]:
+        """Remove duplicates (same stage, different microbatch) and tell me the order i'll encounter the stages in"""
+        seen: Set[int] = set()
+        ret: List[int] = []
+
+        for a in next_actions:
+            if a is not None and a.stage_index not in seen:
+                seen.add(a.stage_index)
+                ret.append(a.stage_index)
+                if len(ret) == count:
+                    break
+        return ret
+
+    active_stages: Set[int] = set()
+    unshard_actions: List[Optional[_Action]] = []
+
+    def _unshard(stage_index: int):
+        active_stages.add(stage_index)
+        unshard_actions.append(_Action(stage_index, UNSHARD))
+
+    def _reshard(stage_index: int):
+        active_stages.remove(stage_index)
+        unshard_actions.append(_Action(stage_index, RESHARD))
+
+    for i, action in enumerate(compute_actions):
+        if action is None:
+            continue
+
+        # We prefetch the next N stages we'll see, dropping existing stages to make room
+        next_n = next_stage_indices(max_active_stages, compute_actions[i:])
+        # Fetch needs to be ordered correctly, so don't use a set
+        fetch = list(filter(lambda s: s not in active_stages, next_n))
+        # Unclear what the best policy is for eviction, but we can maintain order so we do
+        evict = list(filter(lambda s: s not in next_n, active_stages))
+
+        logger.debug(
+            "_add_unshard_reshard Step %d active: %s fetch %s, evict %s",
+            i,
+            active_stages,
+            fetch,
+            evict,
+        )
+
+        for stage in evict:
+            _reshard(stage)
+        for stage in fetch:
+            _unshard(stage)
+        unshard_actions.append(action)
+
+    return unshard_actions
```

Comment:
A general question: how does this interact with the code of FSDP? Which library would be the actual "firer" of the unshard ops?

Reply:
I am working on a PR now that adds the 'runtime' component that would read this schedule format and execute the operations.  Probably what i'll do for the 'unshard' implementation is, 
 - if the stage module is an FSDP instance, call fsdp's unshard
 - otherwise, no-op.  (this way the unshard lowering can be run in all cases even if someone uses it with DDP..) 

If we need to decide at lowering time whether we're using FSDP vs DDP vs no DP, I think we can do that, but it makes composability a little more annoying so i'd rather not.